## Import Python Libraries 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

## Load Data Set
Create a data frame then load the data set. Here the dataset used is WorldCupMatches

In [2]:
#Load data

df = pd.read_csv('WorldCupMatches.csv')

Display the data set that has been loaded

In [3]:
df.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   
3                     2549.0                   1.0                   0.0   
4                    23409.0                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
0         REGO Gilberto (BRA)    201.0   1096.0                FRA   
1       WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2         BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3    MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4  RADULESCU Constantin (ROU)    201.0   1085.0                ARG   

  Away Team Initials  
0                MEX  
1                BEL  
2                BRA  
3                PER  
4                FRA

## Data Cleansing
### Check Data Condition

In [4]:
df.shape

(4572, 20)

Row: 4572
Column:  20

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4572 entries, 0 to 4571
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  852 non-null    float64
 1   Datetime              852 non-null    object 
 2   Stage                 852 non-null    object 
 3   Stadium               852 non-null    object 
 4   City                  852 non-null    object 
 5   Home Team Name        852 non-null    object 
 6   Home Team Goals       852 non-null    float64
 7   Away Team Goals       852 non-null    float64
 8   Away Team Name        852 non-null    object 
 9   Win conditions        852 non-null    object 
 10  Attendance            850 non-null    float64
 11  Half-time Home Goals  852 non-null    float64
 12  Half-time Away Goals  852 non-null    float64
 13  Referee               852 non-null    object 
 14  Assistant 1           852 non-null    object 
 15  Assistant 2          

From the description, we can see that some columns have fewer non-null values than the total entries (4572), indicating the presence of null or empty values in the DataFrame. Additionally, we can also observe the data types of each column, where there are columns whose data types are not yet appropriate.

## Check and remove duplicate rows from theData Frame
### keep
- {‘first’, ‘last’, False}, default ‘first’
- Determines which duplicates (if any) to mark.
- first : Mark duplicates as True except for the first occurrence.
- last : Mark duplicates as True except for the last occurrence.
- False : Mark all duplicates as True.

In [6]:
df[df.duplicated(keep='first')]

Year              Datetime        Stage              Stadium  \
836   2014.0  28 Jun 2014 - 13:00   Round of 16     Estadio Mineirao   
837   2014.0  28 Jun 2014 - 17:00   Round of 16  Estadio do Maracana   
838   2014.0  29 Jun 2014 - 13:00   Round of 16     Estadio Castelao   
839   2014.0  29 Jun 2014 - 17:00   Round of 16     Arena Pernambuco   
840   2014.0  30 Jun 2014 - 13:00   Round of 16     Estadio Nacional   
...      ...                   ...          ...                  ...   
4567     NaN                   NaN          NaN                  NaN   
4568     NaN                   NaN          NaN                  NaN   
4569     NaN                   NaN          NaN                  NaN   
4570     NaN                   NaN          NaN                  NaN   
4571     NaN                   NaN          NaN                  NaN   

                 City Home Team Name  Home Team Goals  Away Team Goals  \
836   Belo Horizonte          Brazil              1.0              1.0   
837   Rio De Janeiro        Colombia              2.0              0.0   
838        Fortaleza     Netherlands              2.0              1.0   
839           Recife      Costa Rica              1.0              1.0   
840         Brasilia          France              2.0              0.0   
...               ...            ...              ...              ...   
4567              NaN            NaN              NaN              NaN   
4568              NaN            NaN              NaN              NaN   
4569              NaN            NaN              NaN              NaN   
4570              NaN            NaN              NaN              NaN   
4571              NaN            NaN              NaN              NaN   

     Away Team Name                        Win conditions  Attendance  \
836           Chile      Brazil win on penalties (3 - 2)      57714.0   
837         Uruguay                                           73804.0   
838          Mexico                                           58817.0   
839          Greece  Costa Rica win on penalties (5 - 3)      41242.0   
840         Nigeria                                           67882.0   
...             ...                                   ...         ...   
4567            NaN                                   NaN         NaN   
4568            NaN                                   NaN         NaN   
4569            NaN                                   NaN         NaN   
4570            NaN                                   NaN         NaN   
4571            NaN                                   NaN         NaN   

      Half-time Home Goals  Half-time Away Goals              Referee  \
836                    0.0                   0.0    WEBB Howard (ENG)   
837                    1.0                   0.0  Bj�rn KUIPERS (NED)   
838                    0.0                   0.0  PROENCA Pedro (POR)   
839                    0.0                   0.0   Ben WILLIAMS (AUS)   
840                    0.0                   0.0    GEIGER Mark (USA)   
...                    ...                   ...                  ...   
4567                   NaN                   NaN                  NaN   
4568                   NaN                   NaN                  NaN   
4569                   NaN                   NaN                  NaN   
4570                   NaN                   NaN                  NaN   
4571                   NaN                   NaN                  NaN   

                  Assistant 1           Assistant 2   RoundID      MatchID  \
836   MULLARKEY Michael (ENG)     Darren CANN (ENG)  255951.0  300186487.0   
837   Sander VAN ROEKEL (NED)  Erwin ZEINSTRA (NED)  255951.0  300186491.0   
838     MIRANDA Bertino (POR)      TRIGO Jose (POR)  255951.0  300186508.0   
839       CREAM Matthew (AUS)      ANAZ Hakan (AUS)  255951.0  300186459.0   
840           HURD Sean (USA)    FLETCHER Joe (CAN)  255951.0  300186462.0   
...                       ...   

In [7]:
df[df.duplicated(keep='last')]

Year              Datetime           Stage              Stadium  \
820   2014.0  28 Jun 2014 - 13:00      Round of 16     Estadio Mineirao   
821   2014.0  28 Jun 2014 - 17:00      Round of 16  Estadio do Maracana   
822   2014.0  30 Jun 2014 - 13:00      Round of 16     Estadio Nacional   
823   2014.0  30 Jun 2014 - 17:00      Round of 16    Estadio Beira-Rio   
824   2014.0  04 Jul 2014 - 17:00   Quarter-finals     Estadio Castelao   
...      ...                   ...             ...                  ...   
4566     NaN                   NaN             NaN                  NaN   
4567     NaN                   NaN             NaN                  NaN   
4568     NaN                   NaN             NaN                  NaN   
4569     NaN                   NaN             NaN                  NaN   
4570     NaN                   NaN             NaN                  NaN   

                 City Home Team Name  Home Team Goals  Away Team Goals  \
820   Belo Horizonte          Brazil              1.0              1.0   
821   Rio De Janeiro        Colombia              2.0              0.0   
822         Brasilia          France              2.0              0.0   
823     Porto Alegre         Germany              2.0              1.0   
824        Fortaleza          Brazil              2.0              1.0   
...               ...            ...              ...              ...   
4566              NaN            NaN              NaN              NaN   
4567              NaN            NaN              NaN              NaN   
4568              NaN            NaN              NaN              NaN   
4569              NaN            NaN              NaN              NaN   
4570              NaN            NaN              NaN              NaN   

     Away Team Name                    Win conditions  Attendance  \
820           Chile  Brazil win on penalties (3 - 2)      57714.0   
821         Uruguay                                       73804.0   
822         Nigeria                                       67882.0   
823         Algeria     Germany win after extra time          NaN   
824        Colombia                                       60342.0   
...             ...                               ...         ...   
4566            NaN                               NaN         NaN   
4567            NaN                               NaN         NaN   
4568            NaN                               NaN         NaN   
4569            NaN                               NaN         NaN   
4570            NaN                               NaN         NaN   

      Half-time Home Goals  Half-time Away Goals  \
820                    0.0                   0.0   
821                    1.0                   0.0   
822                    0.0                   0.0   
823                    0.0                   0.0   
824                    1.0                   0.0   
...                    ...                   ...   
4566                   NaN                   NaN   
4567                   NaN                   NaN   
4568                   NaN                   NaN   
4569                   NaN                   NaN   
4570                   NaN                   NaN   

                            Referee                     Assistant 1  \
820               WEBB Howard (ENG)         MULLARKEY Michael (ENG)   
821             Bj�rn KUIPERS (NED)         Sander VAN ROEKEL (NED)   
822               GEIGER Mark (USA)                 HURD Sean (USA)   
823              RICCI Sandro (BRA)       DE CARVALHO Emerson (BRA)   
824   Carlos VELASCO CARBALLO (ESP)  ALONSO FERNANDEZ Roberto (ESP)   
...                             ...                             ...   
4566                            NaN                             NaN   
4567                            NaN                             NaN   
4568                            NaN                             NaN   
4569                            NaN                   

In [8]:
df[df.duplicated(keep=False)]

Year              Datetime           Stage              Stadium  \
820   2014.0  28 Jun 2014 - 13:00      Round of 16     Estadio Mineirao   
821   2014.0  28 Jun 2014 - 17:00      Round of 16  Estadio do Maracana   
822   2014.0  30 Jun 2014 - 13:00      Round of 16     Estadio Nacional   
823   2014.0  30 Jun 2014 - 17:00      Round of 16    Estadio Beira-Rio   
824   2014.0  04 Jul 2014 - 17:00   Quarter-finals     Estadio Castelao   
...      ...                   ...             ...                  ...   
4567     NaN                   NaN             NaN                  NaN   
4568     NaN                   NaN             NaN                  NaN   
4569     NaN                   NaN             NaN                  NaN   
4570     NaN                   NaN             NaN                  NaN   
4571     NaN                   NaN             NaN                  NaN   

                 City Home Team Name  Home Team Goals  Away Team Goals  \
820   Belo Horizonte          Brazil              1.0              1.0   
821   Rio De Janeiro        Colombia              2.0              0.0   
822         Brasilia          France              2.0              0.0   
823     Porto Alegre         Germany              2.0              1.0   
824        Fortaleza          Brazil              2.0              1.0   
...               ...            ...              ...              ...   
4567              NaN            NaN              NaN              NaN   
4568              NaN            NaN              NaN              NaN   
4569              NaN            NaN              NaN              NaN   
4570              NaN            NaN              NaN              NaN   
4571              NaN            NaN              NaN              NaN   

     Away Team Name                    Win conditions  Attendance  \
820           Chile  Brazil win on penalties (3 - 2)      57714.0   
821         Uruguay                                       73804.0   
822         Nigeria                                       67882.0   
823         Algeria     Germany win after extra time          NaN   
824        Colombia                                       60342.0   
...             ...                               ...         ...   
4567            NaN                               NaN         NaN   
4568            NaN                               NaN         NaN   
4569            NaN                               NaN         NaN   
4570            NaN                               NaN         NaN   
4571            NaN                               NaN         NaN   

      Half-time Home Goals  Half-time Away Goals  \
820                    0.0                   0.0   
821                    1.0                   0.0   
822                    0.0                   0.0   
823                    0.0                   0.0   
824                    1.0                   0.0   
...                    ...                   ...   
4567                   NaN                   NaN   
4568                   NaN                   NaN   
4569                   NaN                   NaN   
4570                   NaN                   NaN   
4571                   NaN                   NaN   

                            Referee                     Assistant 1  \
820               WEBB Howard (ENG)         MULLARKEY Michael (ENG)   
821             Bj�rn KUIPERS (NED)         Sander VAN ROEKEL (NED)   
822               GEIGER Mark (USA)                 HURD Sean (USA)   
823              RICCI Sandro (BRA)       DE CARVALHO Emerson (BRA)   
824   Carlos VELASCO CARBALLO (ESP)  ALONSO FERNANDEZ Roberto (ESP)   
...                             ...                             ...   
4567                            NaN                             NaN   
4568                            NaN                             NaN   
4569                            NaN                             NaN   
4570                            NaN                   

In [9]:
df.duplicated().sum()

3735

Count duplicate is 3735

Remove these duplicate rows to ensure data integrity and accuracy.

Due to the presence of numerous NaN values, our first step will be to drop rows containing NaN values.

In [10]:
df.dropna(inplace=True)

In [11]:
df.duplicated().sum()

15

In [12]:
df[df.duplicated(keep='first')]

Year              Datetime                     Stage  \
836  2014.0  28 Jun 2014 - 13:00                Round of 16   
837  2014.0  28 Jun 2014 - 17:00                Round of 16   
838  2014.0  29 Jun 2014 - 13:00                Round of 16   
839  2014.0  29 Jun 2014 - 17:00                Round of 16   
840  2014.0  30 Jun 2014 - 13:00                Round of 16   
842  2014.0  01 Jul 2014 - 13:00                Round of 16   
843  2014.0  01 Jul 2014 - 17:00                Round of 16   
844  2014.0  04 Jul 2014 - 13:00             Quarter-finals   
845  2014.0  04 Jul 2014 - 17:00             Quarter-finals   
846  2014.0  05 Jul 2014 - 13:00             Quarter-finals   
847  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
848  2014.0  08 Jul 2014 - 17:00                Semi-finals   
849  2014.0  09 Jul 2014 - 17:00                Semi-finals   
850  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
851  2014.0  13 Jul 2014 - 16:00                      Final   

                 Stadium             City Home Team Name  Home Team Goals  \
836     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
837  Estadio do Maracana  Rio De Janeiro        Colombia              2.0   
838     Estadio Castelao       Fortaleza     Netherlands              2.0   
839     Arena Pernambuco          Recife      Costa Rica              1.0   
840     Estadio Nacional        Brasilia          France              2.0   
842   Arena de Sao Paulo       Sao Paulo       Argentina              1.0   
843     Arena Fonte Nova        Salvador         Belgium              2.0   
844  Estadio do Maracana  Rio De Janeiro          France              0.0   
845     Estadio Castelao       Fortaleza          Brazil              2.0   
846     Estadio Nacional        Brasilia       Argentina              1.0   
847     Arena Fonte Nova        Salvador     Netherlands              0.0   
848     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
849   Arena de Sao Paulo       Sao Paulo     Netherlands              0.0   
850     Estadio Nacional        Brasilia          Brazil              0.0   
851  Estadio do Maracana  Rio De Janeiro         Germany              1.0   

     Away Team Goals Away Team Name                         Win conditions  \
836              1.0          Chile       Brazil win on penalties (3 - 2)    
837              0.0        Uruguay                                          
838              1.0         Mexico                                          
839              1.0         Greece   Costa Rica win on penalties (5 - 3)    
840              0.0        Nigeria                                          
842              0.0    Switzerland        Argentina win after extra time    
843              1.0            USA          Belgium win after extra time    
844              1.0        Germany                                          
845              1.0       Colombia                                          
846              0.0        Belgium                                          
847              0.0     Costa Rica  Netherlands win on penalties (4 - 3)    
848              7.0        Germany                                          
849              0.0      Argentina    Argentina win on penalties (2 - 4)    
850              3.0    Netherlands                                          
851              0.0      Argentina          Germany win after extra time    

     Attendance  Half-time Home Goals  Half-time Away Goals  \
836     57714.0                   0.0                   0.0   
837     73804.0                   1.0                   0.0   
838     58817.0                   0.0                   0.0   
839     41242.0                   0.0                   0.0   
840     67882.0                   0.0                   0.0   
842     63255.0                   0.0                   0.0   
843     51227.0                   0.0                   0.0   
844     7424

In [13]:
df[df.duplicated(keep='last')]

Year              Datetime                     Stage  \
820  2014.0  28 Jun 2014 - 13:00                Round of 16   
821  2014.0  28 Jun 2014 - 17:00                Round of 16   
822  2014.0  30 Jun 2014 - 13:00                Round of 16   
824  2014.0  04 Jul 2014 - 17:00             Quarter-finals   
825  2014.0  04 Jul 2014 - 13:00             Quarter-finals   
826  2014.0  08 Jul 2014 - 17:00                Semi-finals   
827  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
828  2014.0  13 Jul 2014 - 16:00                      Final   
829  2014.0  09 Jul 2014 - 17:00                Semi-finals   
830  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
831  2014.0  05 Jul 2014 - 13:00             Quarter-finals   
832  2014.0  29 Jun 2014 - 13:00                Round of 16   
833  2014.0  29 Jun 2014 - 17:00                Round of 16   
834  2014.0  01 Jul 2014 - 13:00                Round of 16   
835  2014.0  01 Jul 2014 - 17:00                Round of 16   

                 Stadium             City Home Team Name  Home Team Goals  \
820     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
821  Estadio do Maracana  Rio De Janeiro        Colombia              2.0   
822     Estadio Nacional        Brasilia          France              2.0   
824     Estadio Castelao       Fortaleza          Brazil              2.0   
825  Estadio do Maracana  Rio De Janeiro          France              0.0   
826     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
827     Estadio Nacional        Brasilia          Brazil              0.0   
828  Estadio do Maracana  Rio De Janeiro         Germany              1.0   
829   Arena de Sao Paulo       Sao Paulo     Netherlands              0.0   
830     Arena Fonte Nova        Salvador     Netherlands              0.0   
831     Estadio Nacional        Brasilia       Argentina              1.0   
832     Estadio Castelao       Fortaleza     Netherlands              2.0   
833     Arena Pernambuco          Recife      Costa Rica              1.0   
834   Arena de Sao Paulo       Sao Paulo       Argentina              1.0   
835     Arena Fonte Nova        Salvador         Belgium              2.0   

     Away Team Goals Away Team Name                         Win conditions  \
820              1.0          Chile       Brazil win on penalties (3 - 2)    
821              0.0        Uruguay                                          
822              0.0        Nigeria                                          
824              1.0       Colombia                                          
825              1.0        Germany                                          
826              7.0        Germany                                          
827              3.0    Netherlands                                          
828              0.0      Argentina          Germany win after extra time    
829              0.0      Argentina    Argentina win on penalties (2 - 4)    
830              0.0     Costa Rica  Netherlands win on penalties (4 - 3)    
831              0.0        Belgium                                          
832              1.0         Mexico                                          
833              1.0         Greece   Costa Rica win on penalties (5 - 3)    
834              0.0    Switzerland        Argentina win after extra time    
835              1.0            USA          Belgium win after extra time    

     Attendance  Half-time Home Goals  Half-time Away Goals  \
820     57714.0                   0.0                   0.0   
821     73804.0                   1.0                   0.0   
822     67882.0                   0.0                   0.0   
824     60342.0                   1.0                   0.0   
825     74240.0                   0.0                   1.0   
826     58141.0                   0.0                   5.0   
827     68034.0                   0.0                   2.0   
828     7473

In [14]:
df[df.duplicated(keep=False)]

Year              Datetime                     Stage  \
820  2014.0  28 Jun 2014 - 13:00                Round of 16   
821  2014.0  28 Jun 2014 - 17:00                Round of 16   
822  2014.0  30 Jun 2014 - 13:00                Round of 16   
824  2014.0  04 Jul 2014 - 17:00             Quarter-finals   
825  2014.0  04 Jul 2014 - 13:00             Quarter-finals   
826  2014.0  08 Jul 2014 - 17:00                Semi-finals   
827  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
828  2014.0  13 Jul 2014 - 16:00                      Final   
829  2014.0  09 Jul 2014 - 17:00                Semi-finals   
830  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
831  2014.0  05 Jul 2014 - 13:00             Quarter-finals   
832  2014.0  29 Jun 2014 - 13:00                Round of 16   
833  2014.0  29 Jun 2014 - 17:00                Round of 16   
834  2014.0  01 Jul 2014 - 13:00                Round of 16   
835  2014.0  01 Jul 2014 - 17:00                Round of 16   
836  2014.0  28 Jun 2014 - 13:00                Round of 16   
837  2014.0  28 Jun 2014 - 17:00                Round of 16   
838  2014.0  29 Jun 2014 - 13:00                Round of 16   
839  2014.0  29 Jun 2014 - 17:00                Round of 16   
840  2014.0  30 Jun 2014 - 13:00                Round of 16   
842  2014.0  01 Jul 2014 - 13:00                Round of 16   
843  2014.0  01 Jul 2014 - 17:00                Round of 16   
844  2014.0  04 Jul 2014 - 13:00             Quarter-finals   
845  2014.0  04 Jul 2014 - 17:00             Quarter-finals   
846  2014.0  05 Jul 2014 - 13:00             Quarter-finals   
847  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
848  2014.0  08 Jul 2014 - 17:00                Semi-finals   
849  2014.0  09 Jul 2014 - 17:00                Semi-finals   
850  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
851  2014.0  13 Jul 2014 - 16:00                      Final   

                 Stadium             City Home Team Name  Home Team Goals  \
820     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
821  Estadio do Maracana  Rio De Janeiro        Colombia              2.0   
822     Estadio Nacional        Brasilia          France              2.0   
824     Estadio Castelao       Fortaleza          Brazil              2.0   
825  Estadio do Maracana  Rio De Janeiro          France              0.0   
826     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
827     Estadio Nacional        Brasilia          Brazil              0.0   
828  Estadio do Maracana  Rio De Janeiro         Germany              1.0   
829   Arena de Sao Paulo       Sao Paulo     Netherlands              0.0   
830     Arena Fonte Nova        Salvador     Netherlands              0.0   
831     Estadio Nacional        Brasilia       Argentina              1.0   
832     Estadio Castelao       Fortaleza     Netherlands              2.0   
833     Arena Pernambuco          Recife      Costa Rica              1.0   
834   Arena de Sao Paulo       Sao Paulo       Argentina              1.0   
835     Arena Fonte Nova        Salvador         Belgium              2.0   
836     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
837  Estadio do Maracana  Rio De Janeiro        Colombia              2.0   
838     Estadio Castelao       Fortaleza     Netherlands              2.0   
839     Arena Pernambuco          Recife      Costa Rica              1.0   
840     Estadio Nacional        Brasilia          France              2.0   
842   Arena de Sao Paulo       Sao Paulo       Argentina              1.0   
843     Arena Fonte Nova        Salvador         Belgium              2.0   
844  Estadio do Maracana  Rio De Janeiro          France              0.0   
845     Estadio Castelao       Fortaleza          Brazil              2.0   
846     Estadio Nacional        Brasilia       Argentina              1.0   
847     Arena Fonte Nova        Salvador     Nether

In [15]:
df.drop_duplicates(inplace=True)

In [16]:
df.duplicated().sum()

0

In [17]:
df.shape  # Check the shape of the DataFrame after removing duplicates

(835, 20)

## Convert Dtype
Converting data types ensures that the data is in the correct format for analysis and allows for more efficient and accurate data processing.

### Year

In [18]:
# Replacing non-finite values with valid ones (e.g., 0)
df['Year'] = df['Year'].fillna(0).astype(int)

In [19]:
# Deleting rows containing non-finite values while changing the data type.
df = df.dropna(subset=['Year'])
df['Year'] = df['Year'].astype(int)

### Datetime

In [20]:
df['Datetime'] = pd.to_datetime(df['Datetime'], errors='coerce')

### Home Team Goals, Away Team Goals, Attendance, 
### Half-time Home Goals, Half-time Away Goals, RoundID, MatchID

In [21]:
# Replacing non-finite values with valid ones (e.g., 0)
df['Home Team Goals'] = df['Home Team Goals'].fillna(0).astype(int)
df['Away Team Goals'] = df['Away Team Goals'].fillna(0).astype(int)
df['Attendance'] = df['Attendance'].fillna(0).astype(int)
df['Half-time Home Goals'] = df['Half-time Home Goals'].fillna(0).astype(int)
df['Half-time Away Goals'] = df['Half-time Away Goals'].fillna(0).astype(int)
df['RoundID'] = df['RoundID'].fillna(0).astype(int)
df['MatchID'] = df['MatchID'].fillna(0).astype(int)

In [22]:
# Deleting rows containing non-finite values while changing the data type.
df = df.dropna(subset=['Home Team Goals'])
df['Home Team Goals'] = df['Home Team Goals'].astype(int)
df = df.dropna(subset=['Away Team Goals'])
df['Away Team Goals'] = df['Away Team Goals'].astype(int)
df = df.dropna(subset=['Attendance'])
df['Attendance'] = df['Attendance'].astype(int)
df = df.dropna(subset=['Half-time Home Goals'])
df['Half-time Home Goals'] = df['Half-time Home Goals'].astype(int)
df = df.dropna(subset=['Half-time Away Goals'])
df['Half-time Away Goals'] = df['Half-time Away Goals'].astype(int)
df = df.dropna(subset=['RoundID'])
df['RoundID'] = df['RoundID'].astype(int)
df = df.dropna(subset=['MatchID'])
df['MatchID'] = df['MatchID'].astype(int)

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 835 entries, 0 to 835
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Year                  835 non-null    int64         
 1   Datetime              825 non-null    datetime64[ns]
 2   Stage                 835 non-null    object        
 3   Stadium               835 non-null    object        
 4   City                  835 non-null    object        
 5   Home Team Name        835 non-null    object        
 6   Home Team Goals       835 non-null    int64         
 7   Away Team Goals       835 non-null    int64         
 8   Away Team Name        835 non-null    object        
 9   Win conditions        835 non-null    object        
 10  Attendance            835 non-null    int64         
 11  Half-time Home Goals  835 non-null    int64         
 12  Half-time Away Goals  835 non-null    int64         
 13  Referee               835

## Check Percentage Missing Values
checking the percentage of missing values provides valuable insights into the data quality and guides decision-making regarding data preprocessing and analysis strategies.

In [24]:
missing_values = df.isnull().sum()

In [25]:
missing_percentage = (missing_values/len(df)) * 100

In [26]:
missing_info = pd.DataFrame({
    'Column' : missing_values.index,
    'Missing Values' : missing_values.values,
    'Missing Percentage' : missing_percentage.values
})

In [27]:
print(missing_info)

                  Column  Missing Values  Missing Percentage
0                   Year               0            0.000000
1               Datetime              10            1.197605
2                  Stage               0            0.000000
3                Stadium               0            0.000000
4                   City               0            0.000000
5         Home Team Name               0            0.000000
6        Home Team Goals               0            0.000000
7        Away Team Goals               0            0.000000
8         Away Team Name               0            0.000000
9         Win conditions               0            0.000000
10            Attendance               0            0.000000
11  Half-time Home Goals               0            0.000000
12  Half-time Away Goals               0            0.000000
13               Referee               0            0.000000
14           Assistant 1               0            0.000000
15           Assistant 2

It turns out that only the 'Datetime' column has missing values. Although the number of missing values is not significant, considering that 'Datetime' is a crucial column for analysis, we have decided to retain the missing values in the 'Datetime' column.

In [28]:
df.Datetime.value_counts()

Datetime
1934-05-27 16:30:00    8
1958-06-15 19:00:00    7
1958-06-11 19:00:00    7
1958-06-08 19:00:00    7
1970-06-07 12:00:00    4
                      ..
1994-06-21 16:00:00    1
1994-06-21 19:30:00    1
1994-06-21 12:30:00    1
1994-06-20 16:00:00    1
2014-07-01 17:00:00    1
Name: count, Length: 591, dtype: int64

In [29]:
df_sorted = df.sort_values(by='Datetime', ascending=True)
print(df_sorted)

     Year            Datetime        Stage                   Stadium  \
0    1930 1930-07-13 15:00:00      Group 1                   Pocitos   
1    1930 1930-07-13 15:00:00      Group 4            Parque Central   
2    1930 1930-07-14 12:45:00      Group 2            Parque Central   
3    1930 1930-07-14 14:50:00      Group 3                   Pocitos   
4    1930 1930-07-15 16:00:00      Group 1            Parque Central   
..    ...                 ...          ...                       ...   
592  2002                 NaT      Group H      Saitama Stadium 2002   
602  2002                 NaT      Group B  Jeonju World Cup Stadium   
605  2002                 NaT      Group G           Kashima Stadium   
613  2002                 NaT      Group A   Suwon World Cup Stadium   
633  2002                 NaT  Round of 16  Jeonju World Cup Stadium   

            City Home Team Name  Home Team Goals  Away Team Goals  \
0    Montevideo          France                4                1 

In [30]:
df.Datetime.value_counts()

Datetime
1934-05-27 16:30:00    8
1958-06-15 19:00:00    7
1958-06-11 19:00:00    7
1958-06-08 19:00:00    7
1970-06-07 12:00:00    4
                      ..
1994-06-21 16:00:00    1
1994-06-21 19:30:00    1
1994-06-21 12:30:00    1
1994-06-20 16:00:00    1
2014-07-01 17:00:00    1
Name: count, Length: 591, dtype: int64

In [31]:
df['Datetime'] = df['Datetime'].fillna(method='backfill')

/var/folders/9v/jr522vfx6hvf7snl2zzlk8r40000gn/T/ipykernel_92043/1218265266.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df['Datetime'] = df['Datetime'].fillna(method='backfill')


In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 835 entries, 0 to 835
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Year                  835 non-null    int64         
 1   Datetime              835 non-null    datetime64[ns]
 2   Stage                 835 non-null    object        
 3   Stadium               835 non-null    object        
 4   City                  835 non-null    object        
 5   Home Team Name        835 non-null    object        
 6   Home Team Goals       835 non-null    int64         
 7   Away Team Goals       835 non-null    int64         
 8   Away Team Name        835 non-null    object        
 9   Win conditions        835 non-null    object        
 10  Attendance            835 non-null    int64         
 11  Half-time Home Goals  835 non-null    int64         
 12  Half-time Away Goals  835 non-null    int64         
 13  Referee               835

### Insights and Considerations
- All columns have missing values, with 'Attendance' having the highest number of missing values, totaling 3722.
- Assuming we want to analyze "team performance in matches," some unimportant columns can be dropped.

#### Extremely Important Columns
- Year: The year of the match.
- Home Team Name: The name of the home team.
- Away Team Name: The name of the away team.
- Home Team Goals: The score of the match and the performance of the home team in scoring goals.
- Away Team Goals: The score of the match and the performance of the away team in scoring goals.
- Attendance: The number of spectators in the match.
- Datetime: The date and time of the match. (to be manipulated later)

#### Possibly Important Columns
- Half-time Home Goals: The number of goals scored by the home team in the first half.
- Half-time Away Goals: The number of goals scored by the away team in the first half.
- City: The city where the match took place.
- Stadium: The name of the stadium where the match took place.
- Stage: The stage or phase of the tournament.
- RoundID: The ID of the tournament round.
- MatchID: The ID of the match.
- Win conditions: The conditions for winning the match.

#### Unimportant Columns
- Referee: The name of the match referee.
- Assistant 1: The name of the first assistant referee.
- Assistant 2: The name of the second assistant referee.
- Home Team Initials: The initials of the home team.
- Away Team Initials: The initials of the away team.

# Data Manipulation
Data manipulation involves making changes to the dataset to prepare it for analysis or to extract valuable insights. This process includes tasks such as cleaning the data, transforming its structure, creating new features, and aggregating information.

## Drop Non-Essential Columns
This step involves removing columns from the dataset that are deemed non-essential for the analysis. Non-essential columns are those that do not contribute significantly to the analysis or do not align with the research objectives. By dropping these columns, we can streamline the dataset and focus only on the most relevant features.

In [33]:
df.drop(['Referee', 'Assistant 1', 'Assistant 2', 'Home Team Initials', 'Away Team Initials'], 
        axis = 1, inplace=True)
df.columns

Index(['Year', 'Datetime', 'Stage', 'Stadium', 'City', 'Home Team Name',
       'Home Team Goals', 'Away Team Goals', 'Away Team Name',
       'Win conditions', 'Attendance', 'Half-time Home Goals',
       'Half-time Away Goals', 'RoundID', 'MatchID'],
      dtype='object')

We drop columns such as 'Referee', 'Assistant 1', 'Assistant 2', 'Home Team Initials', and 'Away Team Initials' as they are not critical for the analysis at hand.

Show dataset info to see if the date and country columns are missing.

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 835 entries, 0 to 835
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Year                  835 non-null    int64         
 1   Datetime              835 non-null    datetime64[ns]
 2   Stage                 835 non-null    object        
 3   Stadium               835 non-null    object        
 4   City                  835 non-null    object        
 5   Home Team Name        835 non-null    object        
 6   Home Team Goals       835 non-null    int64         
 7   Away Team Goals       835 non-null    int64         
 8   Away Team Name        835 non-null    object        
 9   Win conditions        835 non-null    object        
 10  Attendance            835 non-null    int64         
 11  Half-time Home Goals  835 non-null    int64         
 12  Half-time Away Goals  835 non-null    int64         
 13  RoundID               835

In [35]:
df.head()

Year            Datetime    Stage         Stadium         City  \
0  1930 1930-07-13 15:00:00  Group 1         Pocitos  Montevideo    
1  1930 1930-07-13 15:00:00  Group 4  Parque Central  Montevideo    
2  1930 1930-07-14 12:45:00  Group 2  Parque Central  Montevideo    
3  1930 1930-07-14 14:50:00  Group 3         Pocitos  Montevideo    
4  1930 1930-07-15 16:00:00  Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France                4                1         Mexico   
1            USA                3                0        Belgium   
2     Yugoslavia                2                1         Brazil   
3        Romania                3                1           Peru   
4      Argentina                1                0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                       4444                     3                     0   
1                      18346                     2                     0   
2                      24059                     2                     0   
3                       2549                     1                     0   
4                      23409                     0                     0   

   RoundID  MatchID  
0      201     1096  
1      201     1090  
2      201     1093  
3      201     1098  
4      201     1085

### Make Column Draw, Date, and Time

We will perform data manipulation here. Manipulation here does not involve altering data values but rather restructuring the data to make it easier for machines to read.

To determine whether a match ended in a draw or not, we can use the "Home Team Goals" and "Away Team Goals" columns.

In [36]:
# We'll create a new column to indicate whether the match ended in a draw or not.
df['Draw'] = df['Home Team Goals'] == df['Away Team Goals']

# We'll change the values in the "Draw" column to True or False.
df['Draw'] = df['Draw'].astype(str).replace({'True': 'Yes', 'False': 'No'})

# Display the manipulation result
print(df[['Home Team Goals', 'Away Team Goals', 'Draw']].head())

   Home Team Goals  Away Team Goals Draw
0                4                1   No
1                3                0   No
2                2                1   No
3                3                1   No
4                1                0   No


In [37]:
df.head()

Year            Datetime    Stage         Stadium         City  \
0  1930 1930-07-13 15:00:00  Group 1         Pocitos  Montevideo    
1  1930 1930-07-13 15:00:00  Group 4  Parque Central  Montevideo    
2  1930 1930-07-14 12:45:00  Group 2  Parque Central  Montevideo    
3  1930 1930-07-14 14:50:00  Group 3         Pocitos  Montevideo    
4  1930 1930-07-15 16:00:00  Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France                4                1         Mexico   
1            USA                3                0        Belgium   
2     Yugoslavia                2                1         Brazil   
3        Romania                3                1           Peru   
4      Argentina                1                0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                       4444                     3                     0   
1                      18346                     2                     0   
2                      24059                     2                     0   
3                       2549                     1                     0   
4                      23409                     0                     0   

   RoundID  MatchID Draw  
0      201     1096   No  
1      201     1090   No  
2      201     1093   No  
3      201     1098   No  
4      201     1085   No

In [38]:
df['Tanggal'] = df['Datetime'].dt.date
df['Waktu'] = df['Datetime'].dt.time
df.head()

Year            Datetime    Stage         Stadium         City  \
0  1930 1930-07-13 15:00:00  Group 1         Pocitos  Montevideo    
1  1930 1930-07-13 15:00:00  Group 4  Parque Central  Montevideo    
2  1930 1930-07-14 12:45:00  Group 2  Parque Central  Montevideo    
3  1930 1930-07-14 14:50:00  Group 3         Pocitos  Montevideo    
4  1930 1930-07-15 16:00:00  Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France                4                1         Mexico   
1            USA                3                0        Belgium   
2     Yugoslavia                2                1         Brazil   
3        Romania                3                1           Peru   
4      Argentina                1                0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                       4444                     3                     0   
1                      18346                     2                     0   
2                      24059                     2                     0   
3                       2549                     1                     0   
4                      23409                     0                     0   

   RoundID  MatchID Draw     Tanggal     Waktu  
0      201     1096   No  1930-07-13  15:00:00  
1      201     1090   No  1930-07-13  15:00:00  
2      201     1093   No  1930-07-14  12:45:00  
3      201     1098   No  1930-07-14  14:50:00  
4      201     1085   No  1930-07-15  16:00:00

In [39]:
# Changing the date format to "day - month"
df['Tanggal'] = df['Datetime'].dt.strftime('%d - %b')

# Example to display the manipulation result
print(df[['Datetime', 'Tanggal']].head())

             Datetime   Tanggal
0 1930-07-13 15:00:00  13 - Jul
1 1930-07-13 15:00:00  13 - Jul
2 1930-07-14 12:45:00  14 - Jul
3 1930-07-14 14:50:00  14 - Jul
4 1930-07-15 16:00:00  15 - Jul


In [40]:
df.drop(columns=['Datetime'], inplace=True)

In [41]:
df.rename(columns={'Tanggal': 'Date'}, inplace=True)
df.rename(columns={'Waktu': 'Time'}, inplace=True)

In [42]:
df.head()

Year    Stage         Stadium         City Home Team Name  Home Team Goals  \
0  1930  Group 1         Pocitos  Montevideo          France                4   
1  1930  Group 4  Parque Central  Montevideo             USA                3   
2  1930  Group 2  Parque Central  Montevideo      Yugoslavia                2   
3  1930  Group 3         Pocitos  Montevideo         Romania                3   
4  1930  Group 1  Parque Central  Montevideo       Argentina                1   

   Away Team Goals Away Team Name Win conditions  Attendance  \
0                1         Mexico                       4444   
1                0        Belgium                      18346   
2                1         Brazil                      24059   
3                1           Peru                       2549   
4                0         France                      23409   

   Half-time Home Goals  Half-time Away Goals  RoundID  MatchID Draw  \
0                     3                     0      201     1096   No   
1                     2                     0      201     1090   No   
2                     2                     0      201     1093   No   
3                     1                     0      201     1098   No   
4                     0                     0      201     1085   No   

       Date      Time  
0  13 - Jul  15:00:00  
1  13 - Jul  15:00:00  
2  14 - Jul  12:45:00  
3  14 - Jul  14:50:00  
4  15 - Jul  16:00:00

# WEEK 2
## Data Aggregation and Interpretation Metrics
Data aggregation involves combining and summarizing individual data points into a larger dataset, while interpretation metrics are measures used to make sense of aggregated data, providing insights and understanding.

In [43]:
df.describe()

Year  Home Team Goals  Away Team Goals     Attendance  \
count   835.000000       835.000000       835.000000     835.000000   
mean   1984.500599         1.823952         1.021557   44859.457485   
std      22.289859         1.620137         1.072667   23551.107893   
min    1930.000000         0.000000         0.000000    2000.000000   
25%    1970.000000         1.000000         0.000000   29800.000000   
50%    1990.000000         2.000000         1.000000   41000.000000   
75%    2002.000000         3.000000         2.000000   61009.000000   
max    2014.000000        10.000000         7.000000  173850.000000   

       Half-time Home Goals  Half-time Away Goals       RoundID       MatchID  
count            835.000000            835.000000  8.350000e+02  8.350000e+02  
mean               0.719760              0.427545  1.087363e+07  5.648426e+07  
std                0.942231              0.675333  2.753206e+07  1.067645e+08  
min                0.000000              0.000000  2.010000e+02  2.500000e+01  
25%                0.000000              0.000000  2.620000e+02  1.182000e+03  
50%                0.000000              0.000000  3.370000e+02  2.098000e+03  
75%                1.000000              1.000000  2.497220e+05  4.395005e+07  
max                6.000000              5.000000  9.741060e+07  3.001865e+08

### Interpretation: Storytelling
In 1930, the prestigious football competition known as the "World Cup" was held for the first time, marking an event eagerly anticipated worldwide until 2014, the year covered by this dataset. Throughout this period, the World Cup has been a highly anticipated event, with numerous matches played in stadiums across the globe.

Over the years, from 1930 to 2014, the average number of goals scored by the home team was 1.82, with a standard deviation of 1.62, while the visiting team scored an average of 1.02 goals, with a standard deviation of 1.07. The highest number of goals scored by the home team in a single match was 10, compared to 7 goals by the visiting team.

On average, nearly 45,000 spectators attended each match, filling stadiums worldwide to enjoy the games and support their favorite teams. Analysis of halftime performance revealed that the home team had an advantage in scoring with an average of 0.71 goals per half, with a standard deviation of approximately 0.94 goals. The record for the most goals scored by the home team in the first half of a match was 6.

Conversely, the visiting team scored an average of 0.42 goals per half, with a standard deviation of approximately 0.67 goals. The highest-recorded number of goals by the visiting team in the first half was 5.

Overall, the football competition during this period experienced rapid growth, with matches starting in 1930 and reaching their peak in 2014. However, there was variation in the number of matches each year, with the earlier period (1930-1960) featuring around 10 matches per year, while in recent years, the number has surged to over 50 matches per year. This reflects the evolution and changes in the dynamics of football competition over the past few decades.